File: HackerRank/CTCI/MakingAnagrams.cpp
// https://www.hackerrank.com/challenges/ctci-making-anagrams

#include <map>
#include <set>
#include <list>
#include <cmath>
#include <ctime>
#include <deque>
#include <queue>
#include <stack>
#include <string>
#include <bitset>
#include <cstdio>
#include <limits>
#include <vector>
#include <climits>
#include <cstring>
#include <cstdlib>
#include <fstream>
#include <numeric>
#include <sstream>
#include <iostream>
#include <algorithm>
#include <unordered_map>

using namespace std;

map<char, int> countLetters(string s){
    map<char, int> letters;
    string alphabet = "abcdefghijklmnopqrstuvwxyz";
    for(char letter : alphabet){
        letters[letter] = 0;
    }
    for(char letter : s){
        letters[letter]++;
    }
    return letters;
}

int numberNeeded(string a, string b) {
    map<char, int> aCounts, bCounts;
    int totalLetters = 0;
    aCounts = countLetters(a);
    bCounts = countLetters(b);
    for(auto letter : aCounts){
        totalLetters += abs(letter.second - bCounts[letter.first]);
    }
   return totalLetters;
}

int main(){
    string a;
    cin >> a;
    string b;
    cin >> b;
    cout << numberNeeded(a, b) << endl;
    return 0;
}

File: HackerRank/CTCI/making_anagrams.rb
# https://www.hackerrank.com/challenges/ctci-making-anagrams
#!/bin/ruby

def count_letters (s)
    s_counts = Hash.new(0)
    s.each_char{ |letter|
        s_counts[letter] += 1
    }
    return s_counts
end

def find_anagram(a_counts, b_counts)
    total = 0
    alphabet = "abcdefghijklmnopqrstuvwxyz"
    alphabet.each_char{ |letter|
        total += (a_counts[letter] - b_counts[letter]).abs
    }
    return total
end

a = gets.strip
b = gets.strip

a_counts = count_letters(a)
b_counts = count_letters(b)

puts find_anagram(a_counts, b_counts)
